syncthing_gtk/statusicon.py

```diff
@@ -1,93 +1,569 @@
 	
-	def set(self, icon, text=None):
-		# TODO: That text part on ubuntu...
-		if text == None:
-			text = self._text
+	def _get_popupmenu(self):
+		"""
+		@internal
+		"""
+		return self.__popupmenu
+	
+	def _set_visible(self, visible):
+		"""
+		@internal
+		"""
+		pass
+
+	def do_get_property(self, property):
+		if property.name == "active":
+			return self.__active
 		else:
-			self._text = text
-		if self._si == None:
-			self._create_icon(icon, text)
+			raise AttributeError("Unknown property %s" % property.name)
+	
+	def do_set_property(self, property, value):
+		if property.name == "active":
+			self.__active = value
 		else:
-			if THE_HELL and HAS_INDICATOR:
-				self._si.set_icon(icon)
+			raise AttributeError("unknown property %s" % property.name)
+
+
+class StatusIconDummy(StatusIcon):
+	"""
+	Dummy status icon implementation that does nothing
+	"""
+	def __init__(self, *args):
+		StatusIcon.__init__(self, *args)
+		
+		# Pretty unlikely that this will be visible...
+		self.set_property("active", False)
+	
+	def set(self, icon=None, text=None):
+		StatusIcon.set(self, icon, text)
+		
+		self._get_icon(icon)
+		self._get_text(text)
+
+
+class StatusIconGTK3(StatusIcon):
+	"""
+	Gtk.StatusIcon based status icon backend
+	"""
+	def __init__(self, *args):
+		StatusIcon.__init__(self, *args)
+		
+		if IS_UNITY:
+			# Unity fakes SysTray support but actually hides all icons...
+			raise NotImplementedError
+		
+		if IS_KDE:
+			# While the GTK backend works fine on KDE 4, the StatusIconKDE4 backend will achieve better
+			# results and should be available on any standard KDE 4 installation
+			# (since several KDE applications depend on it)
+			raise NotImplementedError
+		
+		self._tray = Gtk.StatusIcon()
+		
+		self._tray.connect("activate", self._on_click)
+		self._tray.connect("popup-menu", self._on_rclick)
+		self._tray.connect("notify::embedded", self._on_embedded_change)
+		
+		self._tray.set_visible(True)
+		self._tray.set_name("syncthing-gtk")
+		self._tray.set_title(self.TRAY_TITLE)
+		
+		# self._tray.is_embedded() must be called asynchronously
+		# See: http://stackoverflow.com/a/6365904/277882
+		GLib.idle_add(self._on_embedded_change)
+	
+	def set(self, icon=None, text=None):
+		StatusIcon.set(self, icon, text)
+		
+		self._tray.set_from_file(self._get_icon(icon))
+		self._tray.set_tooltip_text(self._get_text(text))
+	
+	def _on_embedded_change(self, *args):
+		# Without an icon update at this point GTK might consider the icon embedded and visible even through
+		# it can't actually be seen...
+		self._tray.set_from_file(self._get_icon())
+		
+		# An invisible tray icon will never be embedded but it also should not be replaced
+		# by a fallback icon
+		is_embedded = self._tray.is_embedded() or not self._tray.get_visible()
+		if is_embedded != self.get_property("active"):
+			self.set_property("active", is_embedded)
+	
+	def _on_rclick(self, si, button, time):
+		self._popupmenu.popup(None, None, None, None, button, time)
```

Comment:
Everywhere else is __popupmenu (with two underscores) used. Thanks to that, I'm getting
```
Traceback (most recent call last):
  File "c:\syncthing-gui\syncthing_gtk\statusicon.py", line 246, in _on_rclick
    self._popupmenu.popup(None, None, None, None, button, time)
AttributeError: 'StatusIconGTK3' object has no attribute '_popupmenu'
```
;-)